postgres/src/lib.rs

```diff
@@ -394,14 +401,59 @@ impl StatementCache {
         query: &str,
         types: &[Type],
     ) -> Result<Statement, Error> {
-        match self.get(query, types) {
-            Some(statement) => Ok(statement),
-            None => {
-                let stmt = client.prepare_typed(query, types).await?;
-                self.insert(query, types, stmt.clone());
-                Ok(stmt)
+        let borrowed_key = StatementCacheKey {
+            query: Cow::Borrowed(query),
+            types: Cow::Borrowed(types),
+        };
+        // A cache entry is either present, in which case we can use it or
+        // absent, in which case we use a semaphore to avoid doing lots of
+        // redundant preparation work. We use an Arc<tokio::sync::Semaphore>
+        // because we can't hold the std::sync::RwLock on self.map across the
+        // await point from tokio_postgres::client::Client::prepare_typed.
```

Comment:
I think we may want to drop or rephrase this second sentence. (and similarly update the commit message) There are many reasons to not just hold the RwLock; for example, it's a very coarse lock, and we'd like to be able to prepare multiple distinct statements concurrently. I think discussing alternative implementation strategies in this comment could be confusing.

Here's a suggested revision, providing an overview of the whole process.
```suggestion
        // Each map entry is empty, contains a Semaphore, or contains a prepared
        // Statement. The first call to `prepare_typed()` will insert a
        // `Semaphore`. Then, tasks will wait for a semaphore permit before
        // preparing the query, to ensure that only one task sends a `PREPARE`
        // statement at once. At both steps in the process, the map will be read
        // via a read lock first, then again via a write lock, before updating
        // the entry.
```